dmriprep/utils/vectors.py

```diff
@@ -0,0 +1,568 @@
+    img_sig_var : bool
+        Check that gradient encoding patterns found in the b-values correspond to those found in the signal intensity
+        variance across volumes of the dwi image. Default is True.
+    """
+
+    from dipy.reconst.dki import check_multi_b
+    from sklearn.cluster import MeanShift, estimate_bandwidth
+    check_val = check_multi_b(gtab, n_bvals, non_zero=False, bmag=bmag)
+    if check_val is False:
+        raise ValueError('Insufficient b-values in gradient table based on the number expected')
+
+    # Check that number of image volumes corresponds to the number of gradient encodings.
+    if dwi_file and img_sig_var is True:
+        volume_num = nib.load(dwi_file).shape[-1]
+        if not len(gtab.bvals) == volume_num:
+            raise Exception("Expected %d total image samples but found %d total gradient encoding values", volume_num,
+                            len(gtab.bvals))
+        # Evaluate B0 locations relative to mean signal variation.
+        data = np.array(nib.load(dwi_file).dataobj)
+        signal_means = []
+        for dir in range(data.shape[-1]):
+            signal_means.append(np.mean(data[:, :, :, dir]))
+        mu = np.mean(signal_means)
+        sigma = np.std(signal_means)
+        low_b = mu+3*sigma
+        signal_b0_indices = np.where(signal_means>low_b)[0]
+
+        # Check B0 locations first
+        if not np.allclose(b0_ixs, signal_b0_indices):
+            raise UserWarning('B0 indices in vectors do not correspond to relative high-signal contrast volumes '
+                              'detected in dwi image.')
+
+        # Next check number of unique b encodings (i.e. shells) and their indices
+        X = np.array(signal_means).reshape(-1, 1)
+        bandwidth = estimate_bandwidth(X, quantile=0.1)
+        ms = MeanShift(bandwidth=bandwidth, bin_seeding=True)
+        ms.fit(X)
+        n_clusters_ = len(ms.cluster_centers_)
+        labs, idx = np.unique(ms.labels_, return_index=True)
+        image_idxs = np.sort(idx[::-1])
+        ix_img = []
+        i = -1
+        for val in range(len(ms.labels_)):
+            if val in image_idxs:
+                i = i + 1
+            ix_img.append(labs[i])
+        ix_img = np.array(ix_img)
+
+        ix_vec = []
+        i = -1
+        for val in range(len(gtab.bvals)):
+            if gtab.bvals[val] != gtab.bvals[val-1]:
+                i = i + 1
+            ix_vec.append(i)
+        ix_vec = np.array(ix_vec)
+
+        if n_clusters_ != len(np.unique(gtab.bvals)):
+            raise UserWarning('Number of unique b-values does not correspond to number of unique signal gradient '
+                              'encoding intensities in the dwi image.')
+
+        if np.all(np.isclose(ix_img, ix_vec)) is True:
+            raise UserWarning('Positions of b-value B0\'s and shell(s) do not correspond to signal intensity '
+                              'fluctuation patterns in the dwi image.')
+    return
```

Comment:
Is this the core of this PR?

Do you think it is worth waiting on #26 to be merged and then add this function to it?

Reply:
Yes, this is the PR. Will rebase this addition with the latest merge from #26.